### PR TITLE
A step by step guide on how to use Typescript with Node.js

### LINKED ISSUE #2123: A step by step guide on how to use Typescript with Node.js
We ask that you please be patient as our team works through approving and publishing all articles/tutorials in a timely manner. Allow 1-3 days for a topic to be reviewed and/or approved - allow 3-7 days for an articles to be reviewed and/or published.

Be sure to visit our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for tools, resources, and example articles to go over.

**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**NOTE: We tend to stray away or tend not to publish reviews/comparisons of commercial product offerings.**

*We're looking for the first 2-3 paragraphs of the article that appropriately summarize what your article will be about.*

### Proposed title of article
Your title should be descriptive of the article/tutorial. Be Specific. *Use keyword research to gain your article higher ranking.*

### Introduction paragraph (2-3 paragraphs):
Please write the Introductory paragraph(s), that would be included in your article. We will use this writing snippet to help us assess overall quality before approval. We're looking for the first 2-3 paragraphs of the article that appropriately summarize what your article will be about.
Take your time and write the content as you would intend to get it published. 

### Key takeaways: 
What is Node.js
What is typescript 
Node.js trends 
Typescript trends 
Why they are both popular
A step-by-step guide to integrate and run Node.js and typescript together 
Using Node.js and typescript to build a simple HTTP server 
### References: 
Please list links to any published content/research that you intend to use to support/guide this article.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by Faithdroid**
…with-nodejs/
This fixes #2123
### Article checklist

**ATTENTION: In your PR - add the ISSUE (topic form) # that this PR corresponds to. You can do this by also adding a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.**
      
      
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**If you have not already please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve our overall technical writing.**

#### Preliminary Checks - Formatting and Structure 
- [ ] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [ ] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [ ] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [ ] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [ ] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [ ] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [ ] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [ ] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [ ] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [ ] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [ ] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [ ] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [ ] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**

### CODE REVIEW

File: content/articles/a-step-by-step-guide-on-how-to-use-typescript-with-nodejs/index.md [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+
+Once Node.js is installed in the computer, run `node –v` to confirm if the installation was successful.
+
+Just like you would have done with any application running on the Node.js runtime, create a project folder. Open a command line and change the directory to point to this new project folder. To generate the Node.js package.json file, run `npm init`. This will introduce systematic questions about your project. Answer this prompts the way you would set up the regular Node.js project. Alternatively, run  `npm init -y` and override the `packages.json` file with the default values.
+
+![packages.json file](/engineering-education/a-step-by-step-guide-on-how-to-use-typescript-with-nodejs/package-json-file.png)
+
+### Setting up Typescript
+
+To set up Typescript with Node.js, you need Typescript dependencies provided by the NPM registry. To install the [Typescript compiler package](https://www.npmjs.com/package/typescript), run the following command.
```
Comment (zolomohan) [inline content/articles/a-step-by-step-guide-on-how-to-use-typescript-with-nodejs/index.md:24]:
```suggestion
To set up Typescript with Node.js, you need Typescript dependencies. To install the [Typescript compiler package](https://www.npmjs.com/package/typescript), run the following command:
```

File: content/articles/a-step-by-step-guide-on-how-to-use-typescript-with-nodejs/index.md [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+This file is the Typescript configuration file. It specifies typescript compile options.
+
+![tsconfig.json file](/engineering-education/a-step-by-step-guide-on-how-to-use-typescript-with-nodejs/tsconfig-file.png)
+
+The `tsconfig.json` file has many options. It's good to know when to turn things on and off. TSC reads this file and uses these options to transpile Typescript into Brower readable JavaScript.
+
+Let's edit this config file to include the necessary option for running Typescript.
+
+- `"target": "es6"` - one key thing you need to define is the version of ECMAScript of JavaScript version that the compiler will output. If you need more advanced ECMAScript features, such as using the arrow function, `const`, and `let`, you can choose to use ES6 or a higher version. The target option allows you to use and follow the new JavaScript patterns. Most of today's browsers only understand the es6 version of JavaScript. we need to tell the Typescript compiler to transpile our Typescript code into es6 JavaScript.
```
Comment (zolomohan) [inline content/articles/a-step-by-step-guide-on-how-to-use-typescript-with-nodejs/index.md:53]:
```suggestion
- `"target": "es6"` - One key thing you need to define is the version of JavaScript the compiler will output. If you need more advanced JavaScript features, such as using the arrow functions, `const`, and `let`, you can choose to use ES6 or a higher version. The target option allows you to use and follow the new JavaScript patterns. Most of today's browsers only understand the es6 version of JavaScript. We need to tell the Typescript compiler to transpile our Typescript code into es6 JavaScript.
```

File: content/articles/a-step-by-step-guide-on-how-to-use-typescript-with-nodejs/index.md [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+    "module": "commonjs",                           
+    "outDir": "./dist",                             
+    "rootDir": "./src",                             
+    "strict": true,
+    "moduleResolution": "node",
+    "esModuleInterop": true,                       
+  },
+  "exclude":[
+    "./node_modules"
+]
```
Comment (zolomohan) [inline content/articles/a-step-by-step-guide-on-how-to-use-typescript-with-nodejs/index.md:92]:
```suggestion
  "exclude":[
    "./node_modules"
  ]
```

File: content/articles/a-step-by-step-guide-on-how-to-use-typescript-with-nodejs/index.md [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+}
+```
+
+Feel free to add or change this file as needed to configure the Typescript compiler to your application and environment needs.
+
+### Using Typescript with Node.js
+
+Go ahead and create the folder `src` in your project directory. `src` host `.ts` files as explained earlier. Inside `src`, create an `index.ts` file and start writing some Typescript code. While writing down Typescript, it is advisable to have a Typescript compiler installed in your project's local dependencies. Run `npm install -D typescript` to do so.
```
Comment (zolomohan) [inline content/articles/a-step-by-step-guide-on-how-to-use-typescript-with-nodejs/index.md:100]:
```suggestion
Go ahead and create the folder `src` in your project directory. `src` hosts the Typescript files as explained earlier. Inside `src`, create an `index.ts` file and start writing some Typescript code. While writing down Typescript, it is advisable to have a Typescript compiler installed in your project's local dependencies. Run `npm install -D typescript` to do so.
```

File: content/articles/a-step-by-step-guide-on-how-to-use-typescript-with-nodejs/index.md [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+Let's edit this config file to include the necessary option for running Typescript.
+
+- `"target": "es6"` - one key thing you need to define is the version of ECMAScript of JavaScript version that the compiler will output. If you need more advanced ECMAScript features, such as using the arrow function, `const`, and `let`, you can choose to use ES6 or a higher version. The target option allows you to use and follow the new JavaScript patterns. Most of today's browsers only understand the es6 version of JavaScript. we need to tell the Typescript compiler to transpile our Typescript code into es6 JavaScript.
+
+- `"module": "commonjs"` - this is the JavaScript module formatting system for structuring and organizing JavaScript code. This let's the compiler use module functions such as `require()` and objects such as `module.exports`.
+
+- `"rootDir": "./src"` - a directory where the input Typescript files are stored. I named the folder are `src`.
+
+- `"outDir": "./build"` - this an output directory where the output structure of the compiled JavaScript will be saved. In this case, JavaScript code will be saved in a folder called to `build` in the root directory of the main project.
```
Comment (zolomohan) [inline content/articles/a-step-by-step-guide-on-how-to-use-typescript-with-nodejs/index.md:59]: You have provided the `outDir` to be `build` here but you've used `dist` in the rest of the article. Please keep it consistent.

File: content/articles/a-step-by-step-guide-on-how-to-use-typescript-with-nodejs/index.md [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+   "start": "ts-node ./src/index.ts"
+}
+```
+
+You can now run `npm start` to execute the `index.ts`. And when you delete the dist folder with the compiled `index.js` then still run `npm start`, you will get the same results.
+
+![Ts-node compiler](/engineering-education/a-step-by-step-guide-on-how-to-use-typescript-with-nodejs/ts-node-compiler.png)
+
+### Using Typescript with packages
+The climax of using Typescript and Node.js is the ability to utilize the open-source NPM packages and frameworks. Let's demonstrated how we can use Typescript to start interacting with NPM libraries.
```
Comment (zolomohan) [inline content/articles/a-step-by-step-guide-on-how-to-use-typescript-with-nodejs/index.md:138]:
```suggestion
The climax of using Typescript and Node.js is the ability to utilize the open-source NPM packages and frameworks. Let me demonstrate how we can use Typescript to start interacting with NPM packages.
```

File: content/articles/a-step-by-step-guide-on-how-to-use-typescript-with-nodejs/index.md [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+
+This example will set up Typescript With Express.js. One thing that you should grab is [Express.js](https://www.npmjs.com/package/express) from the NPM registry. So run `npm install express`.
```
Comment (zolomohan) [inline content/articles/a-step-by-step-guide-on-how-to-use-typescript-with-nodejs/index.md:140]:
```suggestion
This example will set up Typescript With Express.js. You should install [Express.js](https://www.npmjs.com/package/express) from the NPM registry. So run `npm install express`.
```

**Review by zolomohan — CHANGES_REQUESTED** [review]
Hi @Faithdroid. Excellent article. Although, I have a few minor suggestions below. Please ping me when you have implemented them. :)

**Comment by Faithdroid** [comment]
Hello @zolomohan, I have updated the suggested contents and added a hero image

**Comment by hectorkambow** [comment]
Thank you v much for your patience as we worked through our Repo re-structuring process. 

The purpose behind this restructure - was to better allow us to handle more volume and allow students to publish more articles and get more involved with the community (with open-source contribution and suggestions). 

With most of the changes coming to completion - we will be sure to get your article published asap. 

Thank you again - and we are looking forward to more amazing articles.

**Comment by hectorkambow** [comment]
plagiarism check done

File: content/articles/a-step-by-step-guide-on-how-to-use-typescript-with-nodejs/index.md [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+node dist/index.js
+``` 
+
+![Results](/engineering-education/a-step-by-step-guide-on-how-to-use-typescript-with-nodejs/sum-results.png)
+
+Running the above commands every single time to compile and run in a development environment can be annoying. To make this process easier, you need to install a package called `ts-node`.
+
+```bash
+npm install -D ts-node
+```
```
Comment (zolomohan) [inline content/articles/a-step-by-step-guide-on-how-to-use-typescript-with-nodejs/index.md:125]: This package requires typescript to be installed in the local project directory.

File: content/articles/a-step-by-step-guide-on-how-to-use-typescript-with-nodejs/index.md [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+  
+  To use it in your project, edit the `package.json` script tag as follows:
+  
+  ```json
+  "scripts": {
+    "dev": "nodemon ./src/index.ts"
+  }
+  ```
+
+  Then run `npm run dev`.
```
Comment (zolomohan) [inline content/articles/a-step-by-step-guide-on-how-to-use-typescript-with-nodejs/index.md:240]: Could you briefly explain what's the difference between the two packages?

**Review by zolomohan — CHANGES_REQUESTED** [review]

**Comment by zolomohan** [comment]
@Faithdroid We can move to a final review once you update the article with these minor suggestions.

**Comment by zolomohan** [comment]
@Faithdroid Please ping me via Github/Slack when you update the changes. We can move to the final review.

**Comment by zolomohan** [comment]
@Faithdroid, Please add your bio with a profile picture in `content/authors`. Take a look at how others have done it and follow the structure. Ping me via Slack if you need any assistance.

File: content/articles/a-step-by-step-guide-on-how-to-use-typescript-with-nodejs/index.md (hunk 9 above)
Comment (Faithdroid) [reply content/articles/a-step-by-step-guide-on-how-to-use-typescript-with-nodejs/index.md]: > Could you briefly explain what's the difference between the two packages?

I have added a line explaining Nodemon.

**Comment by Faithdroid** [comment]
> @Faithdroid, Please add your bio with a profile picture in `content/authors`. Take a look at how others have done it and follow the structure. Ping me via Slack if you need any assistance.

I have uploaded both the bio and a profile image @zolomohan

**Comment by hectorkambow** [comment]
@Faithdroid great work - nice tutorial very easy to follow along 🚀 

cc @zolomohan

**Pull request merged**